File: ios/Classes/Widget/FlutterAppInboxListView.swift
//
//  FlutterAppInboxListView.swift
//  exponea
//

import Foundation
import ExponeaSDK

public class FlutterAppInboxListViewFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger
    
    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }
    
    public func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        return FlutterAppInboxListView(frame: frame, viewId: viewId, messenger: messenger)
    }
}

public class FlutterAppInboxListView: NSObject, FlutterPlatformView {
    private let channelName = "com.exponea/AppInboxListView"
    private let methodOnAppInboxItemClicked = "onAppInboxItemClicked"
    
    private let _viewController : UIViewController
    private var channel: FlutterMethodChannel

    public func view() -> UIView {
        return _viewController.view
    }
    
    private let frame: CGRect
    private let viewId: Int64
    private let messenger: FlutterBinaryMessenger
    
    init(frame: CGRect, viewId: Int64, messenger: FlutterBinaryMessenger) {
        self.frame = frame
        self.viewId = viewId
        self.messenger = messenger
        channel = FlutterMethodChannel(name: "\(channelName)/\(viewId)", binaryMessenger: messenger)
        _viewController = ExponeaSDK.Exponea.shared.getAppInboxListViewController()
        let viewController = _viewController as? AppInboxListViewController
        super.init()
        viewController?.onItemClickedOverride = { (messageItem: MessageItem, _: Int) in
            self.onItemClicked(item: messageItem)
        }
    }
    
    private func onItemClicked(item: MessageItem) {
        DispatchQueue.main.async {
            self.channel.invokeMethod(self.methodOnAppInboxItemClicked, arguments: try? AppInboxCoder.encode(item))
        }
    }
}

File: android/src/main/kotlin/com/exponea/widget/FlutterAppInboxListView.kt
package com.exponea.widget

import android.content.Context
import android.util.Log
import android.view.View
import com.exponea.data.AppInboxCoder
import com.exponea.sdk.Exponea
import com.exponea.sdk.models.MessageItem
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class FlutterAppInboxListView(
    private val context: Context,
    id: Int,
    private val binding: FlutterPlugin.FlutterPluginBinding,
) : PlatformView {

    private val CHANNEL_NAME = "com.exponea/AppInboxListView"
    private val METHOD_ON_APP_INBOX_ITEM_CLICKED = "onAppInboxItemClicked"
    private var channel: MethodChannel? = null
    private val appInboxListView: View?

    class Factory(private val binding: FlutterPlugin.FlutterPluginBinding) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
        override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
            return FlutterAppInboxListView(context!!, viewId, binding)
        }
    }

    init {
        channel = MethodChannel(binding.binaryMessenger, "$CHANNEL_NAME/$id")
        appInboxListView = Exponea.getAppInboxListView(context = context, onItemClicked = ::onItemClicked)
    }

    private fun onItemClicked(messageItem : MessageItem, index : Int) {
        CoroutineScope(Dispatchers.Main).launch {
            channel?.invokeMethod(
                METHOD_ON_APP_INBOX_ITEM_CLICKED,
                AppInboxCoder.encode(messageItem)
            )
        }
    }

    override fun getView(): View? {
        return appInboxListView
    }

    override fun dispose() {
        channel = null
    }
}